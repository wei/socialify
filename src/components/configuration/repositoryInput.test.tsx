import { fireEvent, render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'

import toast from '@/src/components/toaster'
import RepositoryInput from './repositoryInput'

// Mock dependencies
jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}))

jest.mock('@/src/components/toaster', () => ({
  warning: jest.fn(),
  info: jest.fn(),
}))

const mockPush = jest.fn()
const mockUseRouter = require('next/navigation')
  .useRouter as jest.MockedFunction<any>

describe('RepositoryInput', () => {
  const defaultProps = {
    currentRepository: {
      owner: { login: 'testowner' },
      name: 'testrepo',
    },
    currentSearchParams: 'theme=Light&language=1',
  }

  beforeEach(() => {
    jest.clearAllMocks()
    mockUseRouter.mockReturnValue({
      push: mockPush,
      back: jest.fn(),
      forward: jest.fn(),
      refresh: jest.fn(),
      replace: jest.fn(),
      prefetch: jest.fn(),
    })
  })

  describe('Component Rendering', () => {
    it('renders the repository input component correctly', () => {
      render(<RepositoryInput {...defaultProps} />)

      expect(screen.getByLabelText('Repository')).toBeInTheDocument()
      expect(screen.getByLabelText('GitHub repository')).toBeInTheDocument()
      expect(screen.getByLabelText('Switch to repository')).toBeInTheDocument()
    })

    it('renders with correct styling consistency (snapshot)', () => {
      const { container } = render(<RepositoryInput {...defaultProps} />)

      const input = screen.getByLabelText('GitHub repository')
      const joinContainer = input.closest('.join')
      const submitButton = screen.getByLabelText('Switch to repository')

      // Test input styling snapshot
      expect(input).toMatchSnapshot('repository-input-field')

      // Test container styling for highlight glow and outline snapshot
      expect(joinContainer).toMatchSnapshot('repository-input-container')

      // Test button styling snapshot
      expect(submitButton).toMatchSnapshot('repository-input-submit-button')

      // Test complete component structure snapshot
      expect(container.firstChild).toMatchSnapshot(
        'repository-input-complete-component'
      )
    })

    it('displays GitHub icon correctly (snapshot)', () => {
      render(<RepositoryInput {...defaultProps} />)

      const iconContainer = screen
        .getByLabelText('GitHub repository')
        .closest('.join')
        ?.querySelector('.join-item')
      expect(iconContainer).toBeInTheDocument()
      expect(iconContainer).toMatchSnapshot(
        'repository-input-github-icon-container'
      )
    })
  })

  describe('Input Pre-population', () => {
    it('pre-populates input with current repository value', () => {
      render(<RepositoryInput {...defaultProps} />)

      const input = screen.getByLabelText(
        'GitHub repository'
      ) as HTMLInputElement
      expect(input.value).toBe('testowner/testrepo')
    })

    it('handles different repository formats correctly', () => {
      const propsWithDifferentRepo = {
        ...defaultProps,
        currentRepository: {
          owner: { login: 'different-owner' },
          name: 'different-repo-name',
        },
      }

      render(<RepositoryInput {...propsWithDifferentRepo} />)

      const input = screen.getByLabelText(
        'GitHub repository'
      ) as HTMLInputElement
      expect(input.value).toBe('different-owner/different-repo-name')
    })
  })

  describe('Auto-select on Focus', () => {
    it('auto-selects text when input gains focus', () => {
      render(<RepositoryInput {...defaultProps} />)

      const input = screen.getByLabelText(
        'GitHub repository'
      ) as HTMLInputElement

      // Mock the select method
      const selectSpy = jest.spyOn(input, 'select')

      fireEvent.focus(input)

      expect(selectSpy).toHaveBeenCalled()
    })
  })

  describe('Input Validation', () => {
    it('accepts valid GitHub repository format (owner/repo)', () => {
      render(<RepositoryInput {...defaultProps} />)

      const input = screen.getByLabelText('GitHub repository')
      const submitButton = screen.getByLabelText('Switch to repository')

      fireEvent.change(input, { target: { value: 'validowner/validrepo' } })
      fireEvent.click(submitButton)

      expect(mockPush).toHaveBeenCalledWith(
        '/validowner/validrepo?theme=Light&language=1'
      )
      expect(toast.warning).not.toHaveBeenCalled()
    })

    it('accepts GitHub URLs and extracts owner/repo correctly', () => {
      render(<RepositoryInput {...defaultProps} />)

      const input = screen.getByLabelText('GitHub repository')
      const submitButton = screen.getByLabelText('Switch to repository')

      fireEvent.change(input, {
        target: { value: 'https://github.com/facebook/react' },
      })
      fireEvent.click(submitButton)

      expect(mockPush).toHaveBeenCalledWith(
        '/facebook/react?theme=Light&language=1'
      )
      expect(toast.warning).not.toHaveBeenCalled()
    })

    it('shows warning for invalid repository format', () => {
      render(<RepositoryInput {...defaultProps} />)

      const input = screen.getByLabelText('GitHub repository')
      const submitButton = screen.getByLabelText('Switch to repository')

      fireEvent.change(input, { target: { value: 'invalid-format' } })
      fireEvent.click(submitButton)

      expect(toast.warning).toHaveBeenCalledWith(
        'Please enter a valid GitHub repository.'
      )
      expect(mockPush).not.toHaveBeenCalled()
    })

    it('does not navigate when input is empty or whitespace only', () => {
      render(<RepositoryInput {...defaultProps} />)

      const input = screen.getByLabelText('GitHub repository')
      const submitButton = screen.getByLabelText('Switch to repository')

      fireEvent.change(input, { target: { value: '   ' } }) // whitespace only
      fireEvent.click(submitButton)

      // Should not call navigation or show warning for empty input
      expect(mockPush).not.toHaveBeenCalled()
      expect(toast.warning).not.toHaveBeenCalled()
    })

    it('shows info message when trying to switch to same repository', () => {
      render(<RepositoryInput {...defaultProps} />)

      const submitButton = screen.getByLabelText('Switch to repository')

      // Input is already pre-populated with current repo
      fireEvent.click(submitButton)

      expect(toast.info).toHaveBeenCalledWith(
        'You are already viewing this repository.'
      )
      expect(mockPush).not.toHaveBeenCalled()
    })

    it('handles case-insensitive repository comparison', () => {
      render(<RepositoryInput {...defaultProps} />)

      const input = screen.getByLabelText('GitHub repository')
      const submitButton = screen.getByLabelText('Switch to repository')

      fireEvent.change(input, { target: { value: 'TESTOWNER/TESTREPO' } })
      fireEvent.click(submitButton)

      expect(toast.info).toHaveBeenCalledWith(
        'You are already viewing this repository.'
      )
      expect(mockPush).not.toHaveBeenCalled()
    })
  })

  describe('User Interaction Flows', () => {
    it('handles typing in input field correctly', () => {
      render(<RepositoryInput {...defaultProps} />)

      const input = screen.getByLabelText(
        'GitHub repository'
      ) as HTMLInputElement

      fireEvent.change(input, { target: { value: 'newowner/newrepo' } })

      expect(input.value).toBe('newowner/newrepo')
    })

    it('handles form submission via Enter key', () => {
      render(<RepositoryInput {...defaultProps} />)

      const input = screen.getByLabelText('GitHub repository')
      const form = input.closest('form')!

      fireEvent.change(input, {
        target: { value: 'keyboardowner/keyboardrepo' },
      })
      fireEvent.submit(form)

      expect(mockPush).toHaveBeenCalledWith(
        '/keyboardowner/keyboardrepo?theme=Light&language=1'
      )
    })

    it('preserves search parameters when switching repositories', () => {
      const propsWithParams = {
        ...defaultProps,
        currentSearchParams: 'theme=Dark&font=Inter&language=1&owner=1',
      }

      render(<RepositoryInput {...propsWithParams} />)

      const input = screen.getByLabelText('GitHub repository')
      const submitButton = screen.getByLabelText('Switch to repository')

      fireEvent.change(input, { target: { value: 'newowner/newrepo' } })
      fireEvent.click(submitButton)

      expect(mockPush).toHaveBeenCalledWith(
        '/newowner/newrepo?theme=Dark&font=Inter&language=1&owner=1'
      )
    })

    it('handles empty search parameters correctly', () => {
      const propsWithoutParams = {
        ...defaultProps,
        currentSearchParams: '',
      }

      render(<RepositoryInput {...propsWithoutParams} />)

      const input = screen.getByLabelText('GitHub repository')
      const submitButton = screen.getByLabelText('Switch to repository')

      fireEvent.change(input, { target: { value: 'newowner/newrepo' } })
      fireEvent.click(submitButton)

      expect(mockPush).toHaveBeenCalledWith('/newowner/newrepo')
    })
  })

  describe('Edge Cases', () => {
    it('handles repository names with special characters', () => {
      render(<RepositoryInput {...defaultProps} />)

      const input = screen.getByLabelText('GitHub repository')
      const submitButton = screen.getByLabelText('Switch to repository')

      fireEvent.change(input, {
        target: { value: 'owner/repo-with-dashes_and_underscores.dots' },
      })
      fireEvent.click(submitButton)

      expect(mockPush).toHaveBeenCalledWith(
        '/owner/repo-with-dashes_and_underscores.dots?theme=Light&language=1'
      )
    })

    it('trims whitespace from input', () => {
      render(<RepositoryInput {...defaultProps} />)

      const input = screen.getByLabelText('GitHub repository')
      const submitButton = screen.getByLabelText('Switch to repository')

      fireEvent.change(input, { target: { value: '  owner/repo  ' } })
      fireEvent.click(submitButton)

      expect(mockPush).toHaveBeenCalledWith(
        '/owner/repo?theme=Light&language=1'
      )
    })

    it('handles GitHub URLs with additional path segments', () => {
      render(<RepositoryInput {...defaultProps} />)

      const input = screen.getByLabelText('GitHub repository')
      const submitButton = screen.getByLabelText('Switch to repository')

      fireEvent.change(input, {
        target: { value: 'https://github.com/owner/repo/issues/123' },
      })
      fireEvent.click(submitButton)

      expect(mockPush).toHaveBeenCalledWith(
        '/owner/repo?theme=Light&language=1'
      )
    })
  })
})
