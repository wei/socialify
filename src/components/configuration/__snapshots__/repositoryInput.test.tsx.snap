// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`RepositoryInput Component Rendering displays GitHub icon correctly (snapshot): repository-input-github-icon-container 1`] = `
<div
  class="join-item flex items-center pl-3 bg-base-100 border border-base-300 focus-within:border-primary"
>
  <svg
    aria-hidden="true"
    class="w-4 h-4"
    fill="none"
    height="1em"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M9 19c-5 1.5-5-2.5-7-3m14 6v-3.87a3.37 3.37 0 0 0-.94-2.61c3.14-.35 6.44-1.54 6.44-7A5.44 5.44 0 0 0 20 4.77 5.07 5.07 0 0 0 19.91 1S18.73.65 16 2.48a13.38 13.38 0 0 0-7 0C6.27.65 5.09 1 5.09 1A5.07 5.07 0 0 0 5 4.77a5.44 5.44 0 0 0-1.5 3.78c0 5.42 3.3 6.61 6.44 7A3.37 3.37 0 0 0 9 18.13V22"
    />
  </svg>
</div>
`;

exports[`RepositoryInput Component Rendering renders with correct styling consistency (snapshot): repository-input-complete-component 1`] = `
<div
  class="form-control w-full"
>
  <label
    class="label"
    for="repository-input"
  >
    <span
      class="label-text font-semibold"
    >
      Repository
    </span>
  </label>
  <form>
    <div
      class="join w-full focus-within:outline-none focus-within:ring-2 focus-within:ring-primary focus-within:ring-opacity-50"
    >
      <div
        class="join-item flex items-center pl-3 bg-base-100 border border-base-300 focus-within:border-primary"
      >
        <svg
          aria-hidden="true"
          class="w-4 h-4"
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9 19c-5 1.5-5-2.5-7-3m14 6v-3.87a3.37 3.37 0 0 0-.94-2.61c3.14-.35 6.44-1.54 6.44-7A5.44 5.44 0 0 0 20 4.77 5.07 5.07 0 0 0 19.91 1S18.73.65 16 2.48a13.38 13.38 0 0 0-7 0C6.27.65 5.09 1 5.09 1A5.07 5.07 0 0 0 5 4.77a5.44 5.44 0 0 0-1.5 3.78c0 5.42 3.3 6.61 6.44 7A3.37 3.37 0 0 0 9 18.13V22"
          />
        </svg>
      </div>
      <input
        aria-label="GitHub repository"
        class="join-item input input-sm input-bordered flex-grow text-sm font-bold focus:outline-none border-0 focus:ring-0"
        id="repository-input"
        name="repository-input"
        type="text"
        value="testowner/testrepo"
      />
      <button
        aria-label="Switch to repository"
        class="join-item btn btn-sm px-2 btn-primary flex-shrink-0"
        type="submit"
      >
        <svg
          class="h-4 w-4"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zm-28.9 143.6l75.5 72.4H120c-13.3 0-24 10.7-24 24v16c0 13.3 10.7 24 24 24h182.6l-75.5 72.4c-9.7 9.3-9.9 24.8-.4 34.3l11 10.9c9.4 9.4 24.6 9.4 33.9 0L404.3 273c9.4-9.4 9.4-24.6 0-33.9L271.6 106.3c-9.4-9.4-24.6-9.4-33.9 0l-11 10.9c-9.5 9.6-9.3 25.1.4 34.4z"
          />
        </svg>
      </button>
    </div>
  </form>
</div>
`;

exports[`RepositoryInput Component Rendering renders with correct styling consistency (snapshot): repository-input-container 1`] = `
<div
  class="join w-full focus-within:outline-none focus-within:ring-2 focus-within:ring-primary focus-within:ring-opacity-50"
>
  <div
    class="join-item flex items-center pl-3 bg-base-100 border border-base-300 focus-within:border-primary"
  >
    <svg
      aria-hidden="true"
      class="w-4 h-4"
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9 19c-5 1.5-5-2.5-7-3m14 6v-3.87a3.37 3.37 0 0 0-.94-2.61c3.14-.35 6.44-1.54 6.44-7A5.44 5.44 0 0 0 20 4.77 5.07 5.07 0 0 0 19.91 1S18.73.65 16 2.48a13.38 13.38 0 0 0-7 0C6.27.65 5.09 1 5.09 1A5.07 5.07 0 0 0 5 4.77a5.44 5.44 0 0 0-1.5 3.78c0 5.42 3.3 6.61 6.44 7A3.37 3.37 0 0 0 9 18.13V22"
      />
    </svg>
  </div>
  <input
    aria-label="GitHub repository"
    class="join-item input input-sm input-bordered flex-grow text-sm font-bold focus:outline-none border-0 focus:ring-0"
    id="repository-input"
    name="repository-input"
    type="text"
    value="testowner/testrepo"
  />
  <button
    aria-label="Switch to repository"
    class="join-item btn btn-sm px-2 btn-primary flex-shrink-0"
    type="submit"
  >
    <svg
      class="h-4 w-4"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zm-28.9 143.6l75.5 72.4H120c-13.3 0-24 10.7-24 24v16c0 13.3 10.7 24 24 24h182.6l-75.5 72.4c-9.7 9.3-9.9 24.8-.4 34.3l11 10.9c9.4 9.4 24.6 9.4 33.9 0L404.3 273c9.4-9.4 9.4-24.6 0-33.9L271.6 106.3c-9.4-9.4-24.6-9.4-33.9 0l-11 10.9c-9.5 9.6-9.3 25.1.4 34.4z"
      />
    </svg>
  </button>
</div>
`;

exports[`RepositoryInput Component Rendering renders with correct styling consistency (snapshot): repository-input-field 1`] = `
<input
  aria-label="GitHub repository"
  class="join-item input input-sm input-bordered flex-grow text-sm font-bold focus:outline-none border-0 focus:ring-0"
  id="repository-input"
  name="repository-input"
  type="text"
  value="testowner/testrepo"
/>
`;

exports[`RepositoryInput Component Rendering renders with correct styling consistency (snapshot): repository-input-submit-button 1`] = `
<button
  aria-label="Switch to repository"
  class="join-item btn btn-sm px-2 btn-primary flex-shrink-0"
  type="submit"
>
  <svg
    class="h-4 w-4"
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 512 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zm-28.9 143.6l75.5 72.4H120c-13.3 0-24 10.7-24 24v16c0 13.3 10.7 24 24 24h182.6l-75.5 72.4c-9.7 9.3-9.9 24.8-.4 34.3l11 10.9c9.4 9.4 24.6 9.4 33.9 0L404.3 273c9.4-9.4 9.4-24.6 0-33.9L271.6 106.3c-9.4-9.4-24.6-9.4-33.9 0l-11 10.9c-9.5 9.6-9.3 25.1.4 34.4z"
    />
  </svg>
</button>
`;
